### ISSUE TITLE
Combining @EnableBinding and @RefreshScope breaks channel binding

### ISSUE THREAD

**Issue by styjdt** [spring-cloud/spring-cloud-stream#461]
I've two custom Spring Cloud Stream modules, which communicate with each other via MessageChannels. Basically it is one source and one sink.
The source is implemented in a class annotated with `@EnableBinding(Source.class)` with an `@Autowired` field for taking the `Source` instance to be able to send messages to its output channel.
The sink is implemented in a class annotated with `@EnableBinding(Sink.class)` having one message handler method annotated with `@ServiceActivator(inputChannel = Sink.INPUT)`.

This works as expected, so far so good.

However, when adding the `@RefreshScope` annotation to the classes, the channel binding does not work anymore. Observations are the following:

- the `/channels` endpoint vanishes as soon as the `@RefreshScope` annotation is added, this is identical for the Source application and the Sink application
- adding `@RefreshScope` to only the Source is not that problematic. The autowired Source field is injected anyways, so sending messages to the output channel is still possible
- adding `@RefreshScope` to the Sink causes the `@ServiceActivator`-annotated method never to be called anymore. Instead when a message arrives at the Sink, an exception appears on the (Sink) console:

Exception:

    2016-03-31 13:54:14.262 ERROR 9924 --- [hannel-adapter1] o.s.c.s.b.r.RedisMessageChannelBinder$1 : Failed to deliver message; retries exhausted; message sent to queue 'ERRORS:app_a-to-app_b.anonymous.bc676e8f-88c8-49f6-81e0-16332b390190' 

	org.springframework.messaging.MessageDeliveryException: Dispatcher has no subscribers for channel 'rdms_AppB:8082.input'.; nested exception is org.springframework.integration.MessageDispatchingException: Dispatcher has no subscribers
		at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:81) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutput(AbstractMessageProducingHandler.java:231) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.handler.AbstractMessageProducingHandler.produceOutput(AbstractMessageProducingHandler.java:154) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutputs(AbstractMessageProducingHandler.java:102) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:105) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:147) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:217) ~[spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
		at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:213) ~[spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
		at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:263) [spring-retry-1.1.2.RELEASE.jar:na]
		at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:168) [spring-retry-1.1.2.RELEASE.jar:na]
		at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1.doSend(RedisMessageChannelBinder.java:213) [spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
		at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
		at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.popMessageAndSend(RedisQueueMessageDrivenEndpoint.java:224) [spring-integration-redis-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.access$300(RedisQueueMessageDrivenEndpoint.java:54) [spring-integration-redis-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint$ListenerTask.run(RedisQueueMessageDrivenEndpoint.java:323) [spring-integration-redis-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55) [spring-integration-core-4.2.4.RELEASE.jar:na]
		at java.lang.Thread.run(Thread.java:812) [na:1.8.0_51]
	Caused by: org.springframework.integration.MessageDispatchingException: Dispatcher has no subscribers
		at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:153) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) ~[spring-integration-core-4.2.4.RELEASE.jar:na]
		... 31 common frames omitted

However, rewriting the sink in such a way, that I explicitly declare a Sink field with `@Autowired`-annotated setter in which I subscribe to sink.input() with a new MessageHandler does the trick. In that case the `@EnableBinding(Sink.class)` is still necessary, but the `@ServiceActivator` annotated method can be thrown away. (Omitting the `@EnableBinding(Sink.class)` too causes no message to be delivered at all, i.e. no `handleMessage` called, not even exception on the console.)

So this at least allows the messages to be delivered. Unfortunately either the `@RefreshScope` does still not work with this approach or it is some referencing-old-values-problem: I have a String field annotated with `@Value("${my.property:some default value}")`, which is referenced in the message handler. After a refresh is forced on the Sink application, there is still the old config value used... 
So in fact at the moment I have no possibility to get new configuration values at runtime when working with Spring Cloud Stream.

Am I doing anything wrong? Or is there some "incompatibility bug" regarding this issue?

Alexander

**Comment by artembilan** [comment]
@styjdt ,

Thank you for the detailed report!

To be honest I'm not familiar with `@RefreshScope`, but looks like that is a problem of the Spring Integration, not Spring Cloud Stream. 
Would you mind to share some simple Spring Boot application with just `@RefreshScope` and `@ServiceActivator`.
From one side I will study what is that, from other side we will have a foundation to seek the issue in the Spring Integration. and fix it, of course!

**Comment by ilayaperumalg** [comment]
It looks like the the `ServiceActivator` handler's event driven consumer endpoint with the `inputChannel` is never getting started when `@RefreshScope` is added to the Sink module definition.

@artembilan To simulate this behaviour, you can just add `@RefreshScope` on a sink app like this:

```
@EnableBinding(Sink.class)
@RefreshScope
public class LogSink {

	private static Logger logger = LoggerFactory.getLogger(LogSink.class);

	@ServiceActivator(inputChannel=Sink.INPUT)
	public void loggerSink(Object payload) {
		logger.info("Received: " + payload);
	}

}
```

and, the spring-cloud-context dependency:

```
<dependency>
                        <groupId>org.springframework.cloud</groupId>
                     <artifactId>spring-cloud-context</artifactId>                    
                     <version>1.1.0.RC1</version>
              </dependency>
```

**Comment by artembilan** [comment]
Thank you @ilayaperumalg !

I wonder if this from the Spring Cloud Manual says anything to you all:

> Note
	`@RefreshScope` works (technically) on an `@Configuration` class, but it might lead to surprising behaviour: e.g. it does not mean that all the `@Beans` defined in that class are themselves `@RefreshScope`. Specifically, anything that depends on those beans cannot rely on them being updated when a refresh is initiated, unless it is itself in `@RefreshScope` (in which it will be rebuilt on a refresh and its dependencies re-injected, at which point they will be re-initialized from the refreshed `@Configuration`). 

I think I'll play with that in the nearest future...

**Comment by artembilan** [comment]
I moved a bit further. It doesn't start even without Spring Integration!

@RefreshScope
@Configuration
public class MyConfig implements SmartLifecycle {

    @Override
    public boolean isAutoStartup() {
        return true;
    }

    @Override
    public void start() {
        System.out.println("STARTED");
    }

    @Override
    public void stop() {

    }

    @Override
    public void stop(Runnable callback) {

    }

    @Override
    public boolean isRunning() {
        return false;
    }

    @Override
    public int getPhase() {
        return 0;
    }

}

I wonder if it is know issue in Spring Cloud.
What I have noticed from the Spring Integration perspective that the @Configuration bean comes to the MessagingAnnotationPostProcessor later, than DefaultLifecycleProcessor.start() does its stuff.

So, it doesn't relate to Spring Integration, neither to Spring Cloud Stream.
That is something around @Scope logic.

CC @dsyer

**Comment by artembilan** [comment]
Yep!

We instantiate that bean only here RefreshScope:

@EventListener
public void start(ContextRefreshedEvent event) {
    if (event.getApplicationContext() == this.context) {
        if (this.eager && this.registry != null) {
            for (String name : this.context.getBeanDefinitionNames()) {
                BeanDefinition definition = this.registry.getBeanDefinition(name);
                if (this.getName().equals(definition.getScope())
                        && !definition.isLazyInit()) {
                    this.context.getBean(name);
                }
            }
        }
    }
}

What happens already after the DefaultLifecycleProcessor.start() :smile:

**Comment by artembilan** [comment]
See `AbstractApplicationContext`:
````java
protected void finishRefresh() {
	// Initialize lifecycle processor for this context.
	initLifecycleProcessor();

	// Propagate refresh to lifecycle processor first.
	getLifecycleProcessor().onRefresh();

	// Publish the final event.
	publishEvent(new ContextRefreshedEvent(this));

	// Participate in LiveBeansView MBean, if active.
	LiveBeansView.registerApplicationContext(this);
}
````
Not sure how to fix that yet, but at least here is a bell.

**Comment by mbogoevici** [comment]
Also look at: http://stackoverflow.com/questions/36093390/how-to-use-refreshscope-with-enablebinding-annotated-classes. My suggestion there was to declare the `ServiceActivator` as a separate bean, rather than a method on the `@EnableBinding`-annotated class (which makes it a `@Configuration` and therefore `@Component`).

**Comment by mbogoevici** [comment]
After digging a bit into this, the problem isn't so much with the `@Configuration`, but in the way annotations are processed when `@RefreshScope` comes into play. Basically, `@RefreshScope` beans are by definition lazy and will be instantiated (and postprocessed) only after the context is started (what @artembilan has noticed as well). 

There is a simple way to solve this, by extracting the configuration (i.e. refreshable) properties in a bean of its own, and have that injected into the component. 

```
@SpringBootApplication
@EnableBinding(Source.class)
public class RefreshScopeSourceApplication {

	public static void main(String[] args) {
		SpringApplication.run(RefreshScopeSourceApplication.class,
				"--spring.cloud.stream.bindings.output.destination=mydest"
				);
	}

	@Autowired
	private SourceConfiguration sourceConfiguration;

	@InboundChannelAdapter(Source.OUTPUT)
	public String getMessage() {
		String value = "hello " + sourceConfiguration.getTimestamp();
		System.out.println(value);
		return value;
	}

	@Bean @RefreshScope
	public SourceConfiguration sourceConfiguration() {
		return new SourceConfiguration();
	}
}
```

Where SourceConfiguration could be something like this (probably using `@Value`, but this is for a simple test). 

```
public class SourceConfiguration {

	private long timestamp = System.currentTimeMillis();

	public long getTimestamp() {
		return timestamp;
	}

	public void setTimestamp(long timestamp) {
		this.timestamp = timestamp;
	}
}
```

After refresh, the timestamp value will be updated and messages with the new value will be posted.

**Comment by mbogoevici** [comment]
Closing this as there hasn't been a lot of movement lately - and there is not much to do besides the workaround previously described - if any further action is necessary, let's capture it as a separate issue.

**Comment by fkoner** [comment]
@mbogoevici Works Like a charm:

@styjdt @artembilan @ilayaperumalg  Here is an Example with Properties and @Value

```
import com.google.gson.Gson;
import com.microsip.cloud.orchestration.configuration.InboundTopicConfiguration;
import com.microsip.cloud.orchestration.model.MensajeAgente;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.stream.annotation.EnableBinding;
import org.springframework.cloud.stream.annotation.Input;
import org.springframework.cloud.stream.annotation.StreamListener;
import org.springframework.http.*;
import org.springframework.messaging.SubscribableChannel;
import org.springframework.web.client.RestTemplate;

import java.util.Arrays;

/**
 * Created by paco <paco@technogi.com.mx> on 26/04/17.
 */
@EnableBinding(InboundTopic.InputChannel.class)
public class InboundTopic {
    private static Logger log = LoggerFactory.getLogger(InboundTopic.class);

    @Autowired
    private RestTemplate template;

    @Autowired
    private Gson gson;

    // @see https://github.com/spring-cloud/spring-cloud-stream/issues/461
    // @see http://stackoverflow.com/a/36183203
    @Autowired
    private InboundTopicConfiguration topicConfiguration;

    //@ServiceActivator(inputChannel = InboundTopic.InputChannel.INPUT_CHANNEL)
    @StreamListener(InboundTopic.InputChannel.INPUT_CHANNEL)
    public void processMessage(MensajeAgente mensajeAgente) {
        log.debug("Received: " + mensajeAgente);

        // Recuperamos el Id de Cliente para conocer el tópico al cual se le enviará el mensaje.
        String uri = null;

        HttpHeaders headers = null;
        String request = null;
        HttpEntity<String> entity = null;
        ResponseEntity<String> response = null;

        try {
            uri = (topicConfiguration.getOrchestratorUri().endsWith("/") ? topicConfiguration.getOrchestratorUri().substring(0, topicConfiguration.getOrchestratorUri().length() - 1) : topicConfiguration.getOrchestratorUri())
                    + (topicConfiguration.getTopicPath().startsWith("/") ? "" : "/") + topicConfiguration.getTopicPath() + (topicConfiguration.getTopicPath().endsWith("/") ? "" : "/")
                    + mensajeAgente.getIdCliente();

            headers = new HttpHeaders();
            headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
            headers.setContentType(MediaType.APPLICATION_JSON);

            request = gson.toJson(mensajeAgente);
            entity = new HttpEntity<String>(request, headers);

            log.debug("uri is " + uri);
            log.debug("mensajeAgente is " + mensajeAgente);
            log.debug("request is " + request);
            response = template.exchange(uri, HttpMethod.POST, entity, String.class);

            log.debug("response is " + response);

            if (null != response && HttpStatus.ACCEPTED == response.getStatusCode()) {
                log.debug("Procesado correctamente " + response);
            } else {
                log.debug("No fue posible procesar mensaje " + response);
            }
        } finally {
            //uri = null;
            //headers = null;
            //request = null;
            //entity = null;
            //response = null;
        }
    }

    public interface InputChannel {
        String INPUT_CHANNEL = "mpaas-orchestrator";

        @Input(INPUT_CHANNEL)
        SubscribableChannel inputChannel();
    }
}
```

```
import org.springframework.beans.factory.annotation.Value;

/**
 * Created by paco <paco@technogi.com.mx> on 26/04/17.
 */
public class InboundTopicConfiguration {

    @Value("${com.microsip.cloud.orchestration.server.uri}")
    private String orchestratorUri;

    @Value("${com.microsip.cloud.orchestration.topic.path}")
    private String topicPath;

    public String getOrchestratorUri() {
        return orchestratorUri;
    }

    public void setOrchestratorUri(String orchestratorUri) {
        this.orchestratorUri = orchestratorUri;
    }

    public String getTopicPath() {
        return topicPath;
    }

    public void setTopicPath(String topicPath) {
        this.topicPath = topicPath;
    }
}
```

```
import com.google.gson.Gson;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestTemplate;

/**
 * Created by paco <paco@technogi.com.mx> on 26/04/17.
 */
@Configuration
public class OrchestratorMicroserviceConfiguration {

    @Bean
    public RestTemplate restTemplate() {
        return new RestTemplate();
    }

    @Bean
    public Gson gson() {
        return new Gson();
    }

    // @see https://github.com/spring-cloud/spring-cloud-stream/issues/461
    // @see http://stackoverflow.com/a/36183203
    @Bean @RefreshScope
    public InboundTopicConfiguration inboundTopicConfiguration() {
        return new InboundTopicConfiguration();
    }

}
```
